File: scala/src/test/scala/nonamedb/test/specs/unit/api/ServiceSpec.scala
package nonamedb.test.specs.unit.api

import akka.http.scaladsl.model.StatusCodes
import akka.http.scaladsl.testkit.ScalatestRouteTest
import akka.util.Timeout
import nonamedb.api.Service
import nonamedb.storage.Value
import nonamedb.storage.engines.memory.MemoryEngine
import nonamedb.test.specs.unit.UnitSpec

import scala.concurrent.duration._

class ServiceSpec extends UnitSpec with ScalatestRouteTest {
  "A Service" should "reject invalid requests" in {
    Get() ~> testService.routes ~> check {
      handled should be(false)
    }

    Get("/missing-data") ~> testService.routes ~> check {
      handled should be(false)
    }
  }

  it should "successfully add data" in {
    Post(s"/$testKey").withEntity(testValue) ~> testService.routes ~> check {
      status should be(StatusCodes.OK)
    }
  }

  it should "successfully retrieve new data" in {
    Get(s"/$testKey") ~> testService.routes ~> check {
      status should be(StatusCodes.OK)
      responseAs[Value] should be(testValue)
    }
  }

  it should "successfully update data" in {
    Put(s"/$testKey").withEntity(updatedTestValue) ~> testService.routes ~> check {
      status should be(StatusCodes.OK)
    }
  }

  it should "successfully retrieve updated data" in {
    Get(s"/$testKey") ~> testService.routes ~> check {
      status should be(StatusCodes.OK)
      responseAs[Value] should be(updatedTestValue)
    }
  }

  it should "successfully delete data" in {
    Delete(s"/$testKey") ~> testService.routes ~> check {
      status should be(StatusCodes.OK)
    }
  }

  it should "fail to retrieve delete data" in {
    Get(s"/$testKey") ~> testService.routes ~> check {
      handled should be(false)
    }
  }

  private implicit val timeout: Timeout = 3.seconds
  private val testEngine = new MemoryEngine()
  private val testService = new Service(testEngine)

  private val testKey = "some-key"
  private val testValue = "some value".getBytes
  private val updatedTestValue = "some updated value".getBytes
}

File: kotlin/src/test/kotlin/nonamedb/test/specs/unit/api/ServiceSpec.kt
package nonamedb.test.specs.unit.api

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.http.HttpStatus
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.web.reactive.function.BodyInserters
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToMono
import reactor.test.test

@RunWith(SpringRunner::class)
@SpringBootTest(webEnvironment = RANDOM_PORT)
class ServiceSpec {
    private val testKey = "some-key"
    private val testValue = "some value".toByteArray()
    private val updatedTestValue = "some updated value".toByteArray()

    @LocalServerPort
    var port: Int? = null

    lateinit var client: WebClient

    @Before
    fun setup() {
        client = WebClient.create("http://localhost:$port")
    }

    @Test
    fun shouldRejectInvalidRequests() {
        client.get().uri("/")
                .exchange()
                .test()
                .expectNextMatches { it.statusCode() == HttpStatus.NOT_FOUND }
                .verifyComplete()

        client.get().uri("/missing-data")
                .exchange()
                .test()
                .expectNextMatches { it.statusCode() == HttpStatus.NOT_FOUND }
                .verifyComplete()
    }

    @Test
    fun shouldSuccessfullyRetrieveData() {
        client.post().uri("/$testKey").body(BodyInserters.fromObject(testValue))
                .exchange()
                .test()
                .expectNextMatches { it.statusCode() == HttpStatus.OK }
                .verifyComplete()

        client.get().uri("/$testKey")
                .retrieve()
                .bodyToMono<ByteArray>()
                .test()
                .expectNextMatches { it!!.contentEquals(testValue) }
                .verifyComplete()
    }

    @Test
    fun shouldSuccessfullyRetrieveUpdatedData() {
        client.put().uri("/$testKey").body(BodyInserters.fromObject(updatedTestValue))
                .exchange()
                .test()
                .expectNextMatches { it.statusCode() == HttpStatus.OK }
                .verifyComplete()

        client.get().uri("/$testKey")
                .retrieve()
                .bodyToMono<ByteArray>()
                .test()
                .expectNextMatches { it!!.contentEquals(updatedTestValue) }
                .verifyComplete()
    }

    @Test
    fun shouldFailToRetrieveDeletedData() {
        client.delete().uri("/$testKey")
                .exchange()
                .test()
                .expectNextMatches { it.statusCode() == HttpStatus.OK }
                .verifyComplete()

        client.get().uri("/$testKey")
                .exchange()
                .test()
                .expectNextMatches { it.statusCode() == HttpStatus.NOT_FOUND }
                .verifyComplete()
    }
}
